File: Common/src/main/java/net/hexserver/HexHandler.java
package net.hexserver;

import at.petrak.hexcasting.api.spell.casting.CastingContext;
import at.petrak.hexcasting.api.spell.casting.CastingHarness;
import at.petrak.hexcasting.api.spell.casting.ControllerInfo;
import at.petrak.hexcasting.api.spell.iota.Iota;
import at.petrak.hexcasting.common.lib.hex.HexIotaTypes;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtElement;
import net.minecraft.nbt.StringNbtReader;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Hand;

import java.util.ArrayList;
import java.util.List;

public class HexHandler {
    private static Iota hex;
    public static ServerPlayerEntity player;
    public static ServerWorld world;

    public static List<Iota> castHex(String snbt) {
        if (player == null || world == null) {
            return new ArrayList<>();
        }

        NbtCompound nbt;
        try {
            nbt = StringNbtReader.parse(snbt);
        } catch (CommandSyntaxException e) {
            nbt = new NbtCompound();
        }

        List<NbtElement> nbtList = nbt.getList("hexcasting:data", NbtElement.COMPOUND_TYPE);

        List<Iota> instrs = new ArrayList<>();

        for (NbtElement nbtElement : nbtList) {
            var iota = HexIotaTypes.deserialize((NbtCompound) nbtElement, world);
            instrs.add(iota);
        }

        var sPlayer = player;
        var ctx = new CastingContext(sPlayer, Hand.MAIN_HAND, CastingContext.CastSource.PACKAGED_HEX);
        var harness = new CastingHarness(ctx);
        var result = harness.executeIotas(instrs, sPlayer.getWorld());

        List<Iota> stackResult = new ArrayList<>();
        for (NbtCompound nbtCompound : result.getStack()) {
            var iota = HexIotaTypes.deserialize(nbtCompound, world);
            stackResult.add(iota);
        }

        return stackResult;
    }

    public static List<Iota> debugHex(String snbt) {
        if (player == null || world == null) {
            return new ArrayList<>();
        }

        NbtCompound nbt;
        try {
            nbt = StringNbtReader.parse(snbt);
        } catch (CommandSyntaxException e) {
            nbt = new NbtCompound();
        }

        List<NbtElement> nbtList = nbt.getList("hexcasting:data", NbtElement.COMPOUND_TYPE);

        List<Iota> instrs = new ArrayList<>();

        for (NbtElement nbtElement : nbtList) {
            var iota = HexIotaTypes.deserialize((NbtCompound) nbtElement, world);
            instrs.add(iota);
        }

        var sPlayer = player;
        var ctx = new CastingContext(sPlayer, Hand.MAIN_HAND, CastingContext.CastSource.PACKAGED_HEX);
        var harness = new CastingHarness(ctx);

//        var debugAdapter = DebugAdapterManager[player] ?: return InteractionResultHolder.fail(stack)

        var result = harness.executeIotas(instrs, sPlayer.getWorld());

        List<Iota> stackResult = new ArrayList<>();
        for (NbtCompound nbtCompound : result.getStack()) {
            var iota = HexIotaTypes.deserialize(nbtCompound, world);
            stackResult.add(iota);
        }

        return stackResult;
    }
}

File: Common/src/main/java/net/hexserver/HexHandler.kt
package net.hexserver

import at.petrak.hexcasting.api.spell.casting.CastingContext
import at.petrak.hexcasting.api.spell.casting.CastingContext.CastSource
import at.petrak.hexcasting.api.spell.casting.CastingHarness
import at.petrak.hexcasting.api.spell.iota.Iota
import at.petrak.hexcasting.api.utils.putList
import at.petrak.hexcasting.common.lib.hex.HexIotaTypes
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.sun.net.httpserver.HttpExchange
import gay.`object`.hexdebug.adapter.DebugAdapterManager
import gay.`object`.hexdebug.debugger.CastArgs
import net.hexserver.networking.HexServerNetworking
import net.hexserver.networking.MsgDebugHexC2S
import net.hexserver.networking.MsgRunHexC2S
import net.hexserver.networking.MsgRunHexS2C
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtElement
import net.minecraft.nbt.NbtList
import net.minecraft.nbt.StringNbtReader
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Hand
import java.io.OutputStream


object HexHandlerClient {
    private var exchange: HttpExchange? = null


    fun castHex(snbt: String, response: HttpExchange) {
        this.exchange = response

        val nbt = try {
            StringNbtReader.parse(snbt)
        } catch (e: CommandSyntaxException) {
            NbtCompound()
        }
        HexServerNetworking.sendToServer(MsgRunHexC2S(nbt))
    }

    fun respond(result: List<NbtCompound>) {
        exchange?.also {
            val body = StringBuilder()
            for (iota in result) {
                body.append(HexIotaTypes.getDisplay(iota).string)
                body.append("\n")
            }

            it.sendResponseHeaders(200, body.length.toLong())
            val os: OutputStream = it.responseBody
            os.write(body.toString().toByteArray())
            os.close()

            this.exchange = null
        }
    }

    @Throws(CommandSyntaxException::class)
    fun debugHex(snbt: String) {
        val nbt = StringNbtReader.parse(snbt)

        HexServerNetworking.sendToServer(MsgDebugHexC2S(nbt))
    }
}

object HexHandlerServer {

    fun castHex(nbt: NbtCompound, player: ServerPlayerEntity) {
        val world = player.getWorld()

        val nbtList: List<NbtElement> = nbt.getList("hexcasting:data", NbtElement.COMPOUND_TYPE.toInt())
        val instrs: MutableList<Iota> = ArrayList()
        for (nbtElement in nbtList) {
            val iota = HexIotaTypes.deserialize(nbtElement as NbtCompound, world)
            instrs.add(iota)
        }
        val sPlayer: ServerPlayerEntity = player
        val ctx = CastingContext(sPlayer, Hand.MAIN_HAND, CastSource.PACKAGED_HEX)
        val harness = CastingHarness(ctx)
        val (_, _, stack) = harness.executeIotas(instrs, sPlayer.getWorld())

        val nbtStack = NbtList()
        nbtStack.addAll(stack)

        val nbtResult = NbtCompound()
        nbtResult.putList("cast_result", nbtStack)

        HexServerNetworking.sendToPlayer(player, MsgRunHexS2C(nbtResult))
    }


    fun debugHex(nbt: NbtCompound, player: ServerPlayerEntity) {
        val world = player.getWorld()!!


        val nbtList: List<NbtElement> = nbt.getList("hexcasting:data", NbtElement.COMPOUND_TYPE.toInt())
        val instrs: MutableList<Iota> = ArrayList()
        for (nbtElement in nbtList) {
            val iota = HexIotaTypes.deserialize(nbtElement as NbtCompound, world)
            instrs.add(iota)
        }
        val sPlayer: ServerPlayerEntity = player

        val debugAdapter = DebugAdapterManager[sPlayer]

        val ctx = CastingContext(sPlayer, Hand.MAIN_HAND, CastSource.PACKAGED_HEX)

        val args = CastArgs(instrs, ctx, world) {}

        debugAdapter?.startDebugging(args)
    }
}